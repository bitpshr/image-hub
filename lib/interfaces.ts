export enum API {
	document = "/api/document",
}

export interface UploadDocument {
	byteSize: number;
	date: number;
	filename: string;
}
