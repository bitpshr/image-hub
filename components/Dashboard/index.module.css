@import "~lib/variables.css";

.dashboard {
	position: relative;
}

.count {
	font-size: calc(var(--grid-size) * 3.5);
}

.size {
	margin-bottom: calc(var(--grid-size) * 3.25);
}

.error {
	color: var(--error);
	position: absolute;
	right: 0;
	top: calc(var(--grid-size) * 7.5);
}
