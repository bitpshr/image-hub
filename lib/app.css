@import "https://fonts.googleapis.com/css2?family=Roboto+Mono:wght@100;400;700&display=swap";
@import "~lib/variables.css";

* {
	box-sizing: border-box;
}

html,
body,
#__next {
	min-height: 100%;
}

.image-hub {
	background-color: var(--surface-primary);
	color: var(--text-primary);
	font-family: "Roboto Mono", monospace;
	font-size: var(--font-size-body);
	margin: calc(var(--grid-size) * 8) auto;
	max-width: calc(var(--content-size) + (4 * var(--grid-size)));
	padding: 0 calc(var(--grid-size) * 2);
}
