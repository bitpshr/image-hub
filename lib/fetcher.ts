export default async function fetcher<T = any>(
	url: string,
	init?: RequestInit
): Promise<T> {
	const response = await fetch(url, init);
	if (!response.ok) {
		const errorText = (await response.text()) || "An error has occurred.";
		throw errorText;
	}
	return (!init || init.method !== "DELETE") && response.json();
}
