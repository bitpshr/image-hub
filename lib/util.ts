const formatter = new Intl.NumberFormat("en-US", {
	minimumFractionDigits: 0,
	maximumFractionDigits: 0,
});

export function formatBytes(value: number): string {
	return `${formatter.format(value * 0.001)}Kb`;
}
