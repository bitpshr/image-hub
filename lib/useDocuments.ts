import useSWR from "swr";

import fetcher from "@app/lib/fetcher";
import { API, UploadDocument } from "@app/lib/interfaces";

export interface UseDocumentsValue {
	docs?: UploadDocument[];
	error?: string | number;
	isLoading?: boolean;
}

export default function useDocuments(): UseDocumentsValue {
	const { data, error } = useSWR<UploadDocument[]>(API.document, fetcher);
	return {
		docs: data,
		error: error,
		isLoading: !error && !data,
	};
}
