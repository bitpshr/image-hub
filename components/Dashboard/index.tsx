import { FunctionComponent, memo, useCallback, useMemo, useState } from "react";
import { mutate } from "swr";

import Document from "@app/components/Document";
import Flex from "@app/components/Flex";
import SearchBar from "@app/components/SearchBar";
import fetcher from "@app/lib/fetcher";
import { API } from "@app/lib/interfaces";
import useDocuments from "@app/lib/useDocuments";
import useMedia from "@app/lib/useMedia";
import { formatBytes } from "@app/lib/util";

import css from "./index.module.css";

function hashFile(file: Blob): Promise<string> {
	return new Promise((resolve) => {
		const reader = new FileReader();
		reader.onload = async (event) => {
			if (!event.target) return;
			const bytes = new Uint8Array(event.target.result as ArrayBuffer);
			const hashBuffer = await crypto.subtle.digest("SHA-256", bytes);
			const hashBytes = new Uint8Array(hashBuffer);
			const hash = Array.from(hashBytes)
				.map((b) => b.toString(16).padStart(2, "0"))
				.join("");
			resolve(hash);
		};
		reader.readAsArrayBuffer(file);
	});
}

const Dashboard: FunctionComponent = memo(() => {
	const { docs, isLoading, error: docError } = useDocuments();
	const [search, setSearch] = useState("");
	const [error, setError] = useState(docError);
	const size = useMedia();
	const filteredDocs = useMemo(
		() =>
			docs
				? docs.filter(
						(doc) => doc && doc.filename.toLowerCase().includes(search)
				  )
				: [],
		[docs, search]
	);

	const handleDelete = useCallback((id: string) => {
		const shouldDelete = window.confirm("Are you sure?");
		if (!shouldDelete) return;
		try {
			fetcher(`${API.document}/${id}`, { method: "DELETE" });
			mutate(API.document);
		} catch (error) {
			setError(error);
		}
	}, []);

	const handleSearch = useCallback((event) => {
		setSearch(event.target.value.trim().toLocaleLowerCase());
	}, []);

	const handleUpload = useCallback(async (event) => {
		try {
			const file = event.target.files[0];
			const formData = new FormData();
			formData.append("file", file);
			formData.append("hash", await hashFile(file));
			await fetcher("/api/document", {
				body: formData,
				method: "POST",
			});
			mutate(API.document);
		} catch (error) {
			setError(error);
		}
	}, []);

	return (
		<div className={css.dashboard}>
			<SearchBar onUpload={handleUpload} onChange={handleSearch} />
			{error && <div className={css.error}>{error}</div>}
			{isLoading ? (
				<h1>Loading...</h1>
			) : (
				<Flex alignItems="stretch" direction="column">
					<Flex
						alignItems={size.large ? "flex-end" : "stretch"}
						direction={size.large ? "row" : "column"}
					>
						<Flex grow={1}>
							<h1 className={css.count}>{filteredDocs.length} documents</h1>
						</Flex>
						<Flex className={css.size}>
							Total size:{" "}
							{formatBytes(
								filteredDocs.reduce((prev, curr) => curr.byteSize + prev, 0)
							)}
						</Flex>
					</Flex>
					<Flex grow={1} wrap="wrap">
						{filteredDocs.map((doc) => (
							<Document
								key={doc.filename}
								{...doc}
								onDelete={() => handleDelete(doc.filename)}
							/>
						))}
					</Flex>
				</Flex>
			)}
		</div>
	);
});

export default Dashboard;
