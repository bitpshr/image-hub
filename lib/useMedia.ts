import React from "react";

import { viewportContext } from "@app/components/Viewport";

export interface UseMediaConfig {
	[key: string]: number;
}

export interface UseMediaValue {
	[key: string]: boolean;
}

const BREAKPOINTS = {
	large: 600,
};

export default function useMedia(
	config: UseMediaConfig = BREAKPOINTS
): UseMediaValue {
	const { width } = React.useContext(viewportContext);
	const value: UseMediaValue = {};
	for (const breakpoint in config) {
		value[breakpoint] = !!width && width >= config[breakpoint];
	}
	return value;
}
