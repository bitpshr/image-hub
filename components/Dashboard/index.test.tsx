import "@app/lib/useDocuments";

import React from "react";
import renderer from "react-test-renderer";

import Dashboard from "./";

let mockReturn = {};
jest.mock("@app/lib/useDocuments", () => () => mockReturn);

test("loading state", () => {
	mockReturn = { isLoading: true };
	const component = renderer.create(<Dashboard />);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test("error state", () => {
	mockReturn = { isError: true };
	const component = renderer.create(<Dashboard />);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test("documents", () => {
	mockReturn = {
		documents: [{ filename: "name.png", byteSize: 1337, date: 1000000000 }],
	};
	const component = renderer.create(<Dashboard />);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

afterAll(() => {
	jest.clearAllMocks();
});
