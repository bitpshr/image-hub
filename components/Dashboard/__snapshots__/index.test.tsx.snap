// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`documents 1`] = `
<div
  className="dashboard"
>
  <div
    className="control small"
    style={
      Object {
        "alignItems": "stretch",
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <div
      style={
        Object {
          "alignItems": "flex-start",
          "display": "flex",
          "flexDirection": "row",
          "flexGrow": 1,
        }
      }
    >
      <input
        autoCorrect="off"
        className="search"
        onChange={[Function]}
        placeholder="Search documents..."
        spellCheck={false}
        type="search"
      />
    </div>
    <div
      className="uploadControl"
    >
      <input
        accept="image/x-png,image/jpeg"
        id="upload"
        onChange={[Function]}
        type="file"
      />
      <button
        className="button "
      >
        upload
      </button>
    </div>
  </div>
  <div
    style={
      Object {
        "alignItems": "stretch",
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <div
      style={
        Object {
          "alignItems": "stretch",
          "display": "flex",
          "flexDirection": "column",
        }
      }
    >
      <div
        style={
          Object {
            "alignItems": "flex-start",
            "display": "flex",
            "flexDirection": "row",
            "flexGrow": 1,
          }
        }
      >
        <h1
          className="count"
        >
          0
           documents
        </h1>
      </div>
      <div
        className="size"
        style={
          Object {
            "alignItems": "flex-start",
            "display": "flex",
            "flexDirection": "row",
          }
        }
      >
        Total size:
         
        0Kb
      </div>
    </div>
    <div
      style={
        Object {
          "alignItems": "flex-start",
          "display": "flex",
          "flexDirection": "row",
          "flexGrow": 1,
          "flexWrap": "wrap",
        }
      }
    />
  </div>
</div>
`;

exports[`error state 1`] = `
<div
  className="dashboard"
>
  <div
    className="control small"
    style={
      Object {
        "alignItems": "stretch",
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <div
      style={
        Object {
          "alignItems": "flex-start",
          "display": "flex",
          "flexDirection": "row",
          "flexGrow": 1,
        }
      }
    >
      <input
        autoCorrect="off"
        className="search"
        onChange={[Function]}
        placeholder="Search documents..."
        spellCheck={false}
        type="search"
      />
    </div>
    <div
      className="uploadControl"
    >
      <input
        accept="image/x-png,image/jpeg"
        id="upload"
        onChange={[Function]}
        type="file"
      />
      <button
        className="button "
      >
        upload
      </button>
    </div>
  </div>
  <div
    style={
      Object {
        "alignItems": "stretch",
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <div
      style={
        Object {
          "alignItems": "stretch",
          "display": "flex",
          "flexDirection": "column",
        }
      }
    >
      <div
        style={
          Object {
            "alignItems": "flex-start",
            "display": "flex",
            "flexDirection": "row",
            "flexGrow": 1,
          }
        }
      >
        <h1
          className="count"
        >
          0
           documents
        </h1>
      </div>
      <div
        className="size"
        style={
          Object {
            "alignItems": "flex-start",
            "display": "flex",
            "flexDirection": "row",
          }
        }
      >
        Total size:
         
        0Kb
      </div>
    </div>
    <div
      style={
        Object {
          "alignItems": "flex-start",
          "display": "flex",
          "flexDirection": "row",
          "flexGrow": 1,
          "flexWrap": "wrap",
        }
      }
    />
  </div>
</div>
`;

exports[`loading state 1`] = `
<div
  className="dashboard"
>
  <div
    className="control small"
    style={
      Object {
        "alignItems": "stretch",
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <div
      style={
        Object {
          "alignItems": "flex-start",
          "display": "flex",
          "flexDirection": "row",
          "flexGrow": 1,
        }
      }
    >
      <input
        autoCorrect="off"
        className="search"
        onChange={[Function]}
        placeholder="Search documents..."
        spellCheck={false}
        type="search"
      />
    </div>
    <div
      className="uploadControl"
    >
      <input
        accept="image/x-png,image/jpeg"
        id="upload"
        onChange={[Function]}
        type="file"
      />
      <button
        className="button "
      >
        upload
      </button>
    </div>
  </div>
  <h1>
    Loading...
  </h1>
</div>
`;
